### DISCUSSION TITLE
PutObject failed and Detected unexpected drive ordering refusing to use the drive

### DISCUSSION THREAD

**Discussion by Howie516** [minio/minio#18598] [Q&A] [answered]
## NOTE
If this case is urgent, please subscribe to [Subnet](https://min.io/pricing) so that our 24/7 support team may help you faster.

I cannot using milvus to PutObject to minio, error message in minio logs (all 4 pods) as follows:
`API: SYSTEM()
Time: 14:45:05 UTC 12/05/2023
DeploymentID: 90ef9380-a96b-4c19-bfeb-64228a673905
Error: Write failed. Insufficient number of drives online (*errors.errorString)
       9: internal/logger/logger.go:258:logger.LogIf()
       8: cmd/erasure-encode.go:112:cmd.(*Erasure).Encode()
       7: cmd/erasure-object.go:1190:cmd.erasureObjects.putObject()
       6: cmd/erasure-object.go:954:cmd.erasureObjects.PutObject()
       5: cmd/erasure-sets.go:767:cmd.(*erasureSets).PutObject()
       4: cmd/erasure-server-pool.go:940:cmd.(*erasureServerPools).PutObject()
       3: cmd/config-common.go:83:cmd.saveConfigWithOpts()
       2: cmd/config-common.go:88:cmd.saveConfig()
       1: cmd/data-usage.go:52:cmd.storeDataUsageInBackend()

API: SYSTEM()
Time: 14:45:05 UTC 12/05/2023
DeploymentID: 90ef9380-a96b-4c19-bfeb-64228a673905
Error: Storage resources are insufficient for the write operation .minio.sys/tmp/bda9dc41-3169-4757-9d16-74de2c3e7e24/51714ddc-a7c8-4303-82e3-4fedd6023cd1/part.1 (cmd.InsufficientWriteQuorum)
       3: internal/logger/logger.go:258:logger.LogIf()
       2: cmd/data-scanner.go:228:cmd.runDataScanner()
       1: cmd/data-scanner.go:80:cmd.initDataScanner.func1()

API: SYSTEM()
Time: 14:45:05 UTC 12/05/2023
DeploymentID: 90ef9380-a96b-4c19-bfeb-64228a673905
Error: Storage resources are insufficient for the write operation .minio.sys/tmp/fac3e1f0-6b0c-4f68-bd93-ceb2cf2bcb0d/3ff00182-09e7-4017-93a6-9550bc0b4a19/part.1 (cmd.InsufficientWriteQuorum)
       2: internal/logger/logger.go:258:logger.LogIf()
       1: cmd/data-usage.go:53:cmd.storeDataUsageInBackend()

mc admin trace minio:
`API: PutObject(bucket=milvus-bucket, object=file/insert_log/444833670124586675/444833670124586676/444833670130404652/1/446011509740235315)
Time: 12:06:17 UTC 12/05/2023
DeploymentID: 90ef9380-a96b-4c19-bfeb-64228a673905
RequestID: 179DEE0DDBBC7A71
RemoteHost: x.x.x.x
UserAgent: MinIO (linux; amd64) minio-go/v7.0.56
Error: Write failed. Insufficient number of drives online (*errors.errorString)
       8: internal/logger/logger.go:258:logger.LogIf()
       7: cmd/erasure-encode.go:112:cmd.(*Erasure).Encode()
       6: cmd/erasure-object.go:1190:cmd.erasureObjects.putObject()
       5: cmd/erasure-object.go:954:cmd.erasureObjects.PutObject()
       4: cmd/erasure-sets.go:767:cmd.(*erasureSets).PutObject()
       3: cmd/erasure-server-pool.go:940:cmd.(*erasureServerPools).PutObject()
       2: cmd/object-handlers.go:1865:cmd.objectAPIHandlers.PutObjectHandler()
       1: net/http/server.go:2109:http.HandlerFunc.ServeHTTP()`

## Steps to Reproduce (for bugs)

1. Use k8s to deploy four Pods, one driver per pod
2. Call the minio api to write data in large numbers, 100 threads, each thread to write 10,000 data
3. Observe logs and trace
4.

## Your Environment

* Version used (`minio:RELEASE.2023-03-20T20-16-18Z):
* Server setup and configuration:
* Operating System and version (`uname -a`):

**Comment by ravindk89** [comment, marked as answer]
```
Detected unexpected drive ordering refusing to use the drive: expecting http://milvus-minio-1.milvus-minio-svc.milvus-nmg.svc.cluster.local:9000/export, found http://milvus-minio-3.milvus-minio-svc.milvus-nmg.svc.cluster.local:9000/export, 
```

When MinIO starts, it must track the exact topology of the system. That means it knows which drives are on which nodes.

It knows, for example, that the drive currently mounted on the pod X was previously mounted on pod Y. Therefore it will *not* use that volume, as the order/location of drives matter.

In baremetal environments we see this if users deploy without using `/etc/fstab` and labels, where a restart results in drives being mounted in a different order.

In K8s, it's likely related to the PVC/PV configuration. How have you configured the PV/PVC/Storage Class config and how are they mounted to pods?

**Reply by harshavardhana** [reply]
Host path does not preserve the order of persistent volumes; if you do that, you are on your own.

**Reply by Howie516** [reply]
Each pod has a node specified, ensuring that the pod is mounted to the specified physical path each time it is restarted. So it's not the driver order, is it?

**Reply by Howie516** [reply]
One odd thing is that each pod has the following log. endpoint is minio-3, but the minio-1 path is expected. The endpoint is minio-1, but the path to minio-3 is expected.
Is this related to being responsible for the equilibrium strategy? Can you give us the reasons and solutions？

**Reply by harshavardhana** [reply]
For production issues you have to subscribe @Howie516 https://min.io/pricing

You should let your management know that they should never go production without having MinIO support.

This deployment is not what we recommend, we never recommend using host path based setups.

You can use local PV or our product DirectPV. Hostpath has no node level affinity and it shall not work appropriately in the long run just like the way you faced the problem.

**Comment by harshavardhana** [comment]
> Each pod has a node specified, ensuring that the pod is mounted to the specified physical path each time it is restarted. So it's not the driver order, is it? The configuration file is as follows![image](https://private-user-images.githubusercontent.com/66106785/288273480-dfed5ffc-af7b-4401-a4be-6de1556fd56b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE5MjM2MjUsIm5iZiI6MTcwMTkyMzMyNSwicGF0aCI6Ii82NjEwNjc4NS8yODgyNzM0ODAtZGZlZDVmZmMtYWY3Yi00NDAxLWE0YmUtNmRlMTU1NmZkNTZiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjA3VDA0Mjg0NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUyN2Q5NDdhZmQwMjY2MTdhMmQ0NmU5MzBiNmUyYjQ0MDY1YmE1N2JlYmUyYmNlZWI1NDlhZGU1MjE0ODc3ZGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.t-SPXqQPyULk_5I8tEKqInMBN6f2Fuv7g9oVHPBtT88)

This doesn't do anything when you use hostpath since it has no real affinity.

**Reply by Howie516** [reply]
OK, thanks a lot!  We will try using PVC deployment to fix it

**Comment by beck-8** [comment]
I have a cluster with 80 servers, each with 35 disk slots. Among them, disk21 of node32 was damaged. The computer room replaced it for me. I kept getting an error after formatting and mounting it. I can make sure there are no errors in the path. There should be a problem with data changes in the code. After I restarted the service, the error disappeared. (no disk mounts changed)
```
{"deploymentid":"3ecea552-4e06-4a76-9180-e584e018a05a","level":"ERROR","errKind":"ALL","time":"2023-12-27T01:49:21.79302719Z","api":{"name":"SYSTEM","args":{}},"error":{"message":"Following error has been printed 3 times.. Detected unexpected drive ordering refusing to use the drive: expecting http://node32.gh:9000/mnt/disk35, found http://node32.gh:9000/mnt/disk21, refusing to use the drive (*fmt.wrapError)","source":["internal/logger/logger.go:249:logger.LogAlwaysIf()","cmd/prepare-storage.go:63:cmd.glob..func9.1()","cmd/erasure-sets.go:253:cmd.(*erasureSets).connectDisks.func2()"],"variables":{"endpoint":"http://node32.gh:9000/mnt/disk21"}}}
```
```
root@xx:/webhook# minio -v
minio version RELEASE.2023-12-13T23-28-55Z (commit-id=3781a0f9ad47ed77d1050e1b36848e2ebcf73194)
Runtime: go1.21.5 linux/amd64
License: GNU AGPLv3 <https://www.gnu.org/licenses/agpl-3.0.html>
Copyright: 2015-2023 MinIO, Inc.
```

**Comment by harshavardhana** [comment]
> I have a cluster with 80 servers, each with 35 disk slots. Among them, disk21 of node32 was damaged. The computer room replaced it for me. I kept getting an error after formatting and mounting it. I can make sure there are no errors in the path. There should be a problem with data changes in the code. After I restarted the service, the error disappeared. (no disk mounts changed)
> 
> ```
> {"deploymentid":"3ecea552-4e06-4a76-9180-e584e018a05a","level":"ERROR","errKind":"ALL","time":"2023-12-27T01:49:21.79302719Z","api":{"name":"SYSTEM","args":{}},"error":{"message":"Following error has been printed 3 times.. Detected unexpected drive ordering refusing to use the drive: expecting http://node32.gh:9000/mnt/disk35, found http://node32.gh:9000/mnt/disk21, refusing to use the drive (*fmt.wrapError)","source":["internal/logger/logger.go:249:logger.LogAlwaysIf()","cmd/prepare-storage.go:63:cmd.glob..func9.1()","cmd/erasure-sets.go:253:cmd.(*erasureSets).connectDisks.func2()"],"variables":{"endpoint":"http://node32.gh:9000/mnt/disk21"}}}
> ```
> 
> ```
> root@xx:/webhook# minio -v
> minio version RELEASE.2023-12-13T23-28-55Z (commit-id=3781a0f9ad47ed77d1050e1b36848e2ebcf73194)
> Runtime: go1.21.5 linux/amd64
> License: GNU AGPLv3 <https://www.gnu.org/licenses/agpl-3.0.html>
> Copyright: 2015-2023 MinIO, Inc.
> ```
 

You are getting this on purpose, your drive mapping has changed. Follow the message your disk35 and disk21 needs to be swapped.

This happens because Linux changes ordering of mounts upon reboots at times. Make sure to use 

- labels
- uuids

To keep consistent mounts, we document this as part of production checklist of things to ensure a good production deployment.

I already responded to you for your scale, we require a commercial engagement to help you.

**Reply by beck-8** [reply]
Fully compliant and using LABEL
